keps/sig-release/20190118-creating-packages.md

```diff
@@ -0,0 +1,417 @@
+    used as official releases -- usecase is not clear.
+  - *Usage*:
+    ```
+    k/k # bazel build //build/debs
+    k/k # bazel build //build/rpms
+    ```
+  - *Cons*:
+    - No multi-arch support, builds only `linux/amd64`
+    - rpm packaging needs `rpmbuild` to be installed on the host
+  - *Pros*:
+    - `bazel` integration
+    - picks up proper version from git
+    - package specs and additional files (systemd units, ...) are versioned in 
+  - *Parts to reuse*:
+    - some sort of `bazel` integration
+    - versioning of the packaging specs and additional files alongside the code
+
+[k-release]: https://github.com/kubernetes/release/tree/master/
+[k-k]: https://github.com/kubernetes/kubernetes/tree/master/build/
+[k-release-deb]: https://github.com/kubernetes/release/tree/master/debian
+[k-release-rpm]: https://github.com/kubernetes/release/tree/master/rpm
+[k-k-deb]: https://github.com/kubernetes/kubernetes/tree/master/build/debs
+[k-k-rpm]: https://github.com/kubernetes/kubernetes/tree/master/build/rpms
+
+To a good part this KEP tries to address most the [issue "k/k is the canonical source of all build artefacts"](https://github.com/kubernetes/kubernetes/issues/71677) and related
+issues:
+- [`kubeadm` package should depend on the same version of `kubelet` and
+  `kubectl` packages](https://github.com/kubernetes/kubernetes/issues/72871)
+- [Automate deb/rpm publishing](https://github.com/kubernetes/sig-release/issues/10)
+- ...
+
+
+
+### Goals
+
+- Fold the packaging specs, systemd unit files, dependency configuration, and
+  "Parts to reuse" into one place in `k/k`.
+
+  As a first step, the state from [`k/release/{rpm,debian}`][k-release] should
+  be used, as those are the things we release right now. As a second step all
+  the changes and additions from [`k/k/build/`][k-k] should be merged in.
+
+  There must be **one** package configuration per package, all different
+  packaging formats must use this as source of truth (a package configuration
+  could be *as an example* a config file for a package-generator plus
+  all the needed systemd unit files, ... )
+
+- The tooling must be used for cutting packages in the official release
+  process, it must also be usable in local development to generate one-off
+  packages.
+
+- <a id="pkg-list"></a>The tooling must support packaging
+  - `kubelet`
+  - `kubectl`
+  - `kubeadm`
+  - `cri-tools`
+  - `kubernetes-cni`
+
+  It should be easy to integrate new packages, especially when they are
+  one-binary packages
+
+- <a id="plat-list"></a>The tooling must support packaging for multiple platforms
+  - `linux/amd64`
+  - `linux/arm64`
+  - `linux/armhf`
+  - `linux/ppc64el`
+  - `linux/s390x`
+
+  It should be easy to integrate other platforms, given there is general
+  support by the supported distributions and the cross-building tools.
+  Examples: `kubectl` for `darwin/amd64`, `kubelet` for `windows/386`, ...
```

Comment:
What package format are we going to build for darwin?

Reply:
This is just an example, maybe a stupid one. But maybe we want to ship `kubectl` as a `com.apple.package`. The main intend here is that the currently supported platforms and the supported package formats might change.

Reply:
i wonder if we should keep the main goal / graduation criteria of this KEP to Linux?
...and consider other platforms as extra goals down the implementation timeline.

Reply:
yeah, I think we're going to overcomplicate if we try to solve packaging on all platforms in all possible formats ourselves, but that's just my .02c :^)

Reply:
> i wonder if we should keep the main goal / graduation criteria of this KEP to Linux?

> yeah, I think we're going to overcomplicate if we try to solve packaging on all platforms [...]

I tried to express that with the usage of `must` and `should`. I will mention [RFC 2119][rfc] specifically and groom the KEP for the correct usage of those trigger words.

[rfc]: https://www.ietf.org/rfc/rfc2119.txt

Reply:
SGTM, thanks.
I just wanted to poke at possibly darwin / windows being a bit ambitious in scope after solving linux packaging.

\<sarcasm>
next thing you know k8s might have it's own manager + linux distro 
\</sarcasm> :joy: